### PR TITLE
Job killswitch API

### PR DESCRIPTION
**Pull request by dsoftait**
# What's Changed

- Added `GET` and `POST` `/api/extras/job-results/{id}/revoke/` API endpoints for revoking a running or pending job. GET returns a preview describing whether the job will be terminated (worker still alive) or reaped (worker gone); POST performs the action. Returns 200 on success (preview or revoke), 400 when the strategy reports an error, 403 on permission failure, and 409 when the JobResult is already finished.
- Job revocation (both UI and API) now requires the `extras.run_job` permission.
- All users must have the `extras.run_job` permission to revoke a job. Non-staff users can only revoke jobs they themselves submitted. Staff users can additionally revoke jobs submitted by other users.
- Added `UnknownStrategy` for `RevokeFactory`. Jobs whose `queue_type` is not registered in `RevokeFactory.strategies` now fall back to a reap-only strategy that marks the `JobResult` as revoked without sending a backend signal.
- Added `celery_kwargs` to JobResultFactory
- Added `Revoke Job` button to `JOB_RESULT_BUTTONS`
- Changed tests and docs

# Screenshots
Screenshot 2026-05-14 at 13 29 59
Screenshot 2026-05-14 at 13 30 06

# TODO
- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- n/a Example App Updates (when adding/changing features)
- n/a Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md [hunk 1]
```diff
@@ -26,3 +26,90 @@ Both paths converge on the same final state: a `REVOKED` `JobResult` with all at
...
+
+The button is shown only when the job is in an unfinished state and the current user is permitted to revoke it. See [Permissions](#permissions) for details.
+
+### Revocation via the API
+
+Job revocation can also be triggered via the REST API. The endpoint is exposed on the `JobResult` viewset under `revoke`. Revocation is a two-step flow by design: a first call returns a confirmation preview describing what will happen, and a second call with `confirm=true` performs the action.
+
+#### Preview a revocation
+
+A `POST` without the `confirm` flag (or with `confirm=false`) returns details about the job and the action that would be taken. No worker is signaled and no `JobResult` is modified.
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: I wonder if an alternate (preferred??) pattern might be to have a GET to the REST endpoint return the info and a POST perform the action (whether or not a GET was initially performed), rather than requiring two consecutive POSTs - thoughts?

Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: I was thinking about this as an alternative, so I'm open to suggestions. Would the POST request still require `confirm=True`, or would you like to skip it?

Comment (glennmatthews) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: I don't personally think it's necessary, POSTs by convention are already making changes happen.

File: nautobot/core/tests/test_commands.py [hunk 2]
```diff
@@ -17,7 +19,8 @@ def setUp(self):
         self.user.save()
         self.client.force_login(self.user)
 
-    def test_generate_performance_test_endpoints(self):
+    @mock.patch.object(JobResult, "queue_type", new_callable=mock.PropertyMock, return_value="celery")
```
Comment (glennmatthews) [inline nautobot/core/tests/test_commands.py]: Can you explain why this is needed?

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md [hunk 3]
```diff
@@ -26,3 +26,89 @@ Both paths converge on the same final state: a `REVOKED` `JobResult` with all at
...
+
+## Permissions
+
+Revoking a job requires two things: the right permission and the right relationship to the job.
+
+### Required permission
+
+The action checks `extras.run_job` and `extras.view_jobresult`. The reasoning: revoking is a job-execution concern. A user with broad view access to `JobResult` records should not implicitly gain the ability to terminate live workers. Conversely, a user trusted to run jobs is already trusted to stop their own.
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: This whole addition to the docs feels a bit overly verbose to me. Can we make it more succinct?

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md [hunk 4]
```diff
@@ -26,3 +26,89 @@ Both paths converge on the same final state: a `REVOKED` `JobResult` with all at
...
+
+### Who can revoke which jobs
+
+| User                                  | Can revoke own jobs | Can revoke others' jobs |
+|---------------------------------------|:-------------------:|:-----------------------:|
+| Has `run_job`, not staff              |         Yes         |           No            |
+| Has `run_job`, is staff               |         Yes         |           Yes           |
+| No `run_job`, is staff                |         Yes         |           Yes           |
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: If they're `is_staff` but don't have `run_job`, I would expect "can revoke own jobs" to be "not applicable", rather than "Yes"? Since they shouldn't be able to start any jobs to begin with?

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md [hunk 5]
```diff
@@ -26,3 +26,89 @@ Both paths converge on the same final state: a `REVOKED` `JobResult` with all at
...
+| No `run_job`, not staff               |         No          |           No            |
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: Similarly, is "Can revoke own jobs" "not applicable" in this case?

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md [hunk 6]
```diff
@@ -26,3 +26,89 @@ Both paths converge on the same final state: a `REVOKED` `JobResult` with all at
...
+A `GET` request returns details about the job and the action that would be taken. No worker is signaled and no `JobResult` is modified.
+
+```no-highlight
+curl -X GET \
+-H "Authorization: Token $TOKEN" \
+-H "Accept: application/json; version=1.3; indent=4" \
+http://nautobot/api/extras/job-results/$JOB_RESULT_ID/revoke/
+```
+
+The response indicates whether the job is currently running (terminate path) or whether the worker has gone away (reap path):
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: The below JSON doesn't seem to me to clearly indicate which path will be taken, despite what the above suggests.

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md [hunk 7]
```diff
@@ -26,3 +26,89 @@ Both paths converge on the same final state: a `REVOKED` `JobResult` with all at
...
+-H "Accept: application/json; version=1.3; indent=4" \
+http://nautobot/api/extras/job-results/$JOB_RESULT_ID/revoke/
+```
+
+#### Status codes
+
+| Code | Meaning                                                                                       |
+|------|-----------------------------------------------------------------------------------------------|
+| 200  | Preview returned successfully (`GET`) or revocation succeeded (`POST`).                      |
+| 400  | The revoke strategy reported an error or the queue type is unsupported.                      |
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: Feels like this should be a 5xx code rather than a 400.

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md [hunk 8]
```diff
@@ -26,3 +26,89 @@ Both paths converge on the same final state: a `REVOKED` `JobResult` with all at
...
+| 403  | The caller lacks `extras.run_job` and is not staff, or is not the job owner and is not staff. |
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: We should strongly consider adding a new file in the docs that provides a centralized reference for what all the "is_staff" and "is_superuser" flags control in Nautobot. I think we have an increasing number of such special cases at this point.

File: nautobot/extras/api/views.py [hunk 9]
```diff
@@ -1147,13 +1148,82 @@ class JobResultViewSet(
...
         serializer = serializers.JobLogEntrySerializer(logs, context={"request": request}, many=True)
         return Response(serializer.data)
 
+    @action(detail=True, methods=["get", "post"], permission_classes=[RevokeJobPermission])
+    def revoke(self, request, pk=None):
+        """Terminate a running or pending Job, or reap it if its worker is gone."""
+        job_result = self.get_object()
+
+        if not request.user.has_perm("extras.run_job") and not request.user.is_staff:
+            raise PermissionDenied("Job can not be revoked by user without permission to run jobs.")
```
Comment (glennmatthews) [inline nautobot/extras/api/views.py]: Should we ever check whether the user has permission to run *the specific job associated with the JobResult*? Maybe not, since in the non-is_staff case we only revoke jobs attached to this user, and in the is_staff case we don't check the permission at all - but it feels a bit off?

File: nautobot/extras/api/views.py [hunk 10]
```diff
@@ -1147,13 +1148,82 @@ class JobResultViewSet(
...
+
+        if job_result.user != request.user and not request.user.is_staff:
+            raise PermissionDenied("Job can only be revoked by owners/staff.")
```
Comment (glennmatthews) [inline nautobot/extras/api/views.py]:
```suggestion
            raise PermissionDenied("Job can only be revoked by the submitter or by a staff user.")
```

File: nautobot/extras/views.py [hunk 11]
```diff
@@ -3771,12 +3771,20 @@ def modal(self, request, *args, **kwargs):
...
     def revoke_job(self, request, pk=None):
         """Terminate a running or pending Job, or reap it if its worker is gone."""
         job_result = self.get_object()
 
+        if not request.user.has_perm("extras.run_job") and not request.user.is_staff:
+            messages.error(request, "Job can not be revoked by user without permission to run jobs.")
+            return redirect(job_result.get_absolute_url())
+
+        if job_result.user != request.user and not request.user.is_staff:
+            messages.error(request, "Job can not be revoked only by owners/staff.")
```
Comment (glennmatthews) [inline nautobot/extras/views.py]:
```suggestion
            messages.error(request, "Job can be revoked only by the submitter or by staff users.")
```

File: nautobot/core/tests/test_commands.py (hunk 2 above)
Comment (dsoftait) [reply nautobot/core/tests/test_commands.py]: Sure, our `JobResults` objects in the tests don't include `celery_kwargs` arguments, so when this test calls the `revoke` method from the API, it returns a 400 status code because `queue_type` returns `None`. That's why I had to add this mock. Once we implement this `TODO` from [factory](https://github.com/nautobot/nautobot/blob/2235e34f30eb7741120e56fd29a12681c2e97375/nautobot/extras/factory.py#L200), we’ll be able to remove it.

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 6 above)
Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: hm, we have a `job_status` so in this example is `RUNNING` so we will have a terminate path. Do you have any suggestion how to change it? I tried to follow to similar concept what we already have in UI.

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 8 above)
Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: I agree. So I'm creating a ticket

Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: https://networktocode.atlassian.net/browse/NAUTOBOT-1345

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 4 above)
Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: ok, I see that we have more threads regarding permissions so please let me know how it should work

a) If user `is_staff` then can revoke all jobs without `run_job` permission or
b) If user `is_staff` then can revoke all jobs but have to have `run_job`

File: nautobot/extras/api/views.py (hunk 9 above)
Comment (dsoftait) [reply nautobot/extras/api/views.py]: Ok, so let's define permission in this [thread](https://github.com/nautobot/nautobot/pull/8946/changes#r3227300542)

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 4 above)
Comment (glennmatthews) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: personally I'd lean towards option b, but may need to be a discussion with @itdependsnetworks as well.

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 6 above)
Comment (glennmatthews) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: I was expecting some key in the response that would have an associated value of either "terminate" or "reap", to make it really straightforward to parse.

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 4 above)
Comment (itdependsnetworks) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: Option B makes sense

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 3 above)
Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: changed

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 4 above)
Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: changed

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 5 above)
Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: changed

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 6 above)
Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: added `action` key

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 7 above)
Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: changed

File: nautobot/extras/api/views.py (hunk 9 above)
Comment (dsoftait) [reply nautobot/extras/api/views.py]: changed

File: nautobot/core/tests/test_commands.py (hunk 2 above)
Comment (dsoftait) [reply nautobot/core/tests/test_commands.py]: Added `celery_kwargs` to JobResultFactory

File: nautobot/extras/factory.py [hunk 12]
```diff
@@ -181,7 +181,7 @@ class Params:
         has_user = NautobotBoolIterator(chance_of_getting_true=80)
         has_task_args = NautobotBoolIterator(chance_of_getting_true=10)
         has_task_kwargs = NautobotBoolIterator(chance_of_getting_true=90)
-        # TODO has_scheduled_job? has_meta? has_celery_kwargs?
+        # TODO has_scheduled_job? has_meta?
```
Comment (glennmatthews) [inline nautobot/extras/factory.py:185]: Should we generate some JobResults with empty/missing celery_kwargs still, for negative test purposes?

File: nautobot/extras/api/views.py [hunk 13]
```diff
@@ -1182,35 +1182,34 @@ def revoke(self, request, pk=None):
...
+                ),
                 "job_status": "RUNNING" if job_is_running else "NOT RUNNING",
-                "timestamp": timezone.now().strftime("%Y-%m-%d %H:%M:%S"),
-                "action_description": "This will REAP or TERMINATE the job.",
                 "irreversible": "This action cannot be undone.",
-                "details": {
-                    "REAP": "No worker running; marks JobResult as revoked without signal.",
-                    "TERMINATE": "SIGKILL to worker; stops immediately, no cleanup.",
-                },
+                "timestamp": timezone.now().strftime("%Y-%m-%d %H:%M:%S"),
```
Comment (glennmatthews) [inline nautobot/extras/api/views.py]: Can we use `.isoformat()` instead of `.strftime()`?

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 7 above)
Comment (glennmatthews) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: Doesn't look like the actual view code was changed to return a 500?

File: nautobot/extras/api/views.py (hunk 9 above)
Comment (glennmatthews) [reply nautobot/extras/api/views.py]: with the changed permission, particularly for staff users, I do think we should check to make sure the user has permission to run *the specific job in question* and not allow them to revoke it otherwise. What do you think, @itdependsnetworks, does that seem logical?

File: nautobot/extras/api/views.py [hunk 14]
```diff
@@ -1147,13 +1148,81 @@ class JobResultViewSet(
...
+            super().restrict_queryset(request, *args, **kwargs)
+
     @action(detail=True)
     def logs(self, request, pk=None):
         job_result = self.get_object()
         logs = job_result.job_log_entries.all()
         serializer = serializers.JobLogEntrySerializer(logs, context={"request": request}, many=True)
         return Response(serializer.data)
 
+    @action(detail=True, methods=["get", "post"], permission_classes=[RevokeJobPermission])
```
Comment (glennmatthews) [inline nautobot/extras/api/views.py:1196]: For purposes of the OpenAPI schema and Swagger UI, we ideally should also have an `@extend_schema()` here documenting the request and response format for both GET and POST. Not strictly necessary to define a Serializer class for the GET response but it might be the easiest way to do this.

File: nautobot/extras/tests/test_api.py [hunk 15]
```diff
@@ -3701,6 +3702,174 @@ def setUpTestData(cls):
...
+        )
+        url = reverse("extras-api:jobresult-revoke", kwargs={"pk": self.pending_job_result.pk})
+        response = self.client.post(url, **self.header)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data["id"], str(self.pending_job_result.pk))
+        mock_revoke.assert_called_once()
+
+    @mock.patch.object(CeleryStrategy, "is_alive", return_value=True)
+    @mock.patch.object(CeleryStrategy, "revoke", return_value={"error": "Revoke failed: worker not responding."})
+    def test_revoke_strategy_error_returns_400(self, mock_revoke, mock_is_alive):
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_api.py]: This test case also presumably should be updated from 400 to 500.

Comment (dsoftait) [reply nautobot/extras/tests/test_api.py]: 500

File: nautobot/extras/api/views.py (hunk 9 above)
Comment (dsoftait) [reply nautobot/extras/api/views.py]: confirmed at meeting that current implementation is good

File: nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md (hunk 7 above)
Comment (dsoftait) [reply nautobot/docs/user-guide/platform-functionality/jobs/job-revocation.md]: now should be good

File: nautobot/extras/factory.py (hunk 12 above)
Comment (dsoftait) [reply nautobot/extras/factory.py:185]: added

File: nautobot/extras/api/views.py (hunk 13 above)
Comment (dsoftait) [reply nautobot/extras/api/views.py]: done

File: nautobot/extras/api/views.py (hunk 14 above)
Comment (dsoftait) [reply nautobot/extras/api/views.py:1196]: added

File: nautobot/extras/tests/test_api.py (hunk 15 above)
Comment (dsoftait) [reply nautobot/extras/tests/test_api.py]: done

File: nautobot/extras/jobs_revoke.py [hunk 16]
```diff
@@ -114,21 +114,21 @@ def revoke(self, job_result, user) -> dict:
                 level_choice=LogLevelChoices.LOG_FAILURE,
                 grouping="revoking",
             )
-            return {"job_result": self._mark_revoked(job_result, user), "error": None}
+            return {"job_result": self._mark_revoked(job_result, user), "error": None, "revoked": True}
 
         # TERMINATE
         try:
-            self.perform_termination(job_result, user)
+            revoked = self.perform_termination(job_result, user)
```
Comment (dsoftait) [inline nautobot/extras/jobs_revoke.py:121]: I added this because in perform_termination we check if job is in unready state so can be situation when we don't revoke. Previous I wanted based on job_result.status, but was situation when status wasn't already updated outside this function and information to the user was returned incorrect. That's why I decided to add this flag.

File: nautobot/extras/api/views.py [hunk 17]
```diff
@@ -1209,18 +1225,22 @@ def revoke(self, request, pk=None):
...
 
-        job_result.refresh_from_db()
-        serializer = self.get_serializer(job_result)
+        if not result["revoked"]:
+            return Response(
+                {"detail": "Job finished before it could be revoked. No action was taken."},
+                status=status.HTTP_409_CONFLICT,
+            )
 
+        serializer = self.get_serializer(result["job_result"])
```
Comment (dsoftait) [inline nautobot/extras/api/views.py:1243]: Celery worker set status `REVOKED` when action is "terminate" so then can be situation that user will get job result with not updated status. In UI we update this by htmx, but in API we don't have that kind of tool. Do we want to somehow inform the user about it?

Comment (glennmatthews) [reply nautobot/extras/api/views.py:1243]: I think reporting the status as it is *now* is the best we can do in the API. Maybe just something we need to document clearly.

File: nautobot/extras/jobs_revoke.py [hunk 18]
```diff
@@ -227,12 +227,61 @@ def perform_termination(self, job_result: JobResult, user: User):
...
+
+        The is_unready_state check prevents a task that
+        has already been completed from being overwritten.
+        """
+        # TBD: maybe this method shouldn't be abstract and should be defined in ABC class
+        # Review after implementing K8sStrategy
+        return job_result.is_unready_state and not self.is_alive(job_result)
+
+    def perform_termination(self, job_result: JobResult, user: User) -> bool:
+        """Defensive fallback; normally unreachable.
```
Comment (glennmatthews) [inline nautobot/extras/jobs_revoke.py]: Can we add any test code that exercises this fallback? I'm not sure I fully understand the need for this.

File: nautobot/extras/tables.py [hunk 19]
```diff
@@ -175,6 +175,16 @@
             </a>
         </li>
     {% endif %}
+    {% if record.is_unready_state %}
+        {% if record.user == request.user or request.user.is_staff %}
```
Comment (glennmatthews) [inline nautobot/extras/tables.py:179]: Should we also check for `perms.extras.run_job`?

Comment (dsoftait) [reply nautobot/extras/tables.py:179]: it's already in one if above
https://github.com/nautobot/nautobot/pull/8946/changes#diff-f807218a521768366c9dea654100684163410328fd1493f6110f0aa35cb1c401R155

**Review by glennmatthews — APPROVED** [review]

File: nautobot/extras/tests/test_jobs.py [hunk 20]
```diff
@@ -2319,10 +2319,79 @@ def test_factory_returns_celery_strategy_for_celery_queue(self):
...
+                job_result = self._make_job_result(status)
+                self.assertFalse(strategy.should_reap(job_result))
+
+    def test_unknown_perform_terminate_false_always(self):
+        """perform_termination return always False."""
+        strategy = UnknownStrategy()
+        for status in (*JobResultStatusChoices.UNREADY_STATES, *JobResultStatusChoices.READY_STATES):
+            with self.subTest(status=status):
+                job_result = self._make_job_result(status)
+                self.assertFalse(strategy.perform_termination(job_result))
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_jobs.py]:
```suggestion
                self.assertFalse(strategy.perform_termination(job_result, self.user))
```
